### PR TITLE
Remove the IForegroundNotificationService

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to 

https://github.com/dotnet/roslyn/pull/52793 and
https://github.com/dotnet/roslyn/pull/52812

Removes this service (except for a stub interface to keep binary compat with Typescript).

### CODE REVIEW

File: src/EditorFeatures/Core/Tagging/AsynchronousViewTaggerProvider.cs [hunk 1]
```diff
@@ -21,16 +19,7 @@ protected AsynchronousViewTaggerProvider(IThreadingContext threadingContext, IAs
...
         }
 
-        // TypeScript still is moving to calling the new constructor that does not take a IForegroundNotificationService.
-        // Until then, we can fetch one from another service of ours that already does. When TypeScript moves calling the
-        // new constructor, this should be deleted.
-        [Obsolete("This overload exists for TypeScript compatibility only and should not be used in new code.")]
-        protected AsynchronousViewTaggerProvider(IThreadingContext threadingContext, IAsynchronousOperationListener asyncListener, IForegroundNotificationService _)
-            : base(threadingContext, asyncListener)
-        {
-        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AsynchronousViewTaggerProvider.cs:31]: this constructor is no longer used by TS.  we moved them onto the constructor they should be using.

**Comment by CyrusNajmabadi** [comment]
@sharwell this is ready for review.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs [hunk 2]
```diff
@@ -66,21 +65,35 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
...
+                {
+                    // Keep firing events for this doc, as long as we haven't exceeded the max amount
+                    // of waiting time, and there's no user input that should take precedence.
+                    if (stopwatch.Elapsed.Ticks > MaxTimeSlice || IsInputPending())
+                    {
+                        await Task.Delay(DelayBetweenProcessing, cancellationToken).ConfigureAwait(true);
+                        stopwatch = SharedStopwatch.StartNew();
+                    }
+                }
+                while (FireEventsForDocument(documentId));
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:93]: @sharwell this should actually give the same semantics as before.  However, if we're worried about this, we could keep IFGService, but just move it into the CodeModel project.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs [hunk 3]
```diff
@@ -66,21 +65,35 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
...
-                    () => FireEventsForDocument(documentId),
-                    _listener.BeginAsyncOperation("CodeModelEvent"),
-                    cancellationToken);
+                do
+                {
+                    // Keep firing events for this doc, as long as we haven't exceeded the max amount
+                    // of waiting time, and there's no user input that should take precedence.
+                    if (stopwatch.Elapsed.Ticks > MaxTimeSlice || IsInputPending())
+                    {
+                        await Task.Delay(DelayBetweenProcessing, cancellationToken).ConfigureAwait(true);
```
Comment (sharwell) [inline src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:89]: 📝 This needs to use `IExpeditedDelaySource` (typically `IAsynchronousOperationListener`)

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs [hunk 4]
```diff
@@ -41,14 +41,13 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
             VisualStudioWorkspace visualStudioWorkspace,
             [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
             IThreadingContext threadingContext,
-            IForegroundNotificationService notificationService,
             IAsynchronousOperationListenerProvider listenerProvider)
+            : base(threadingContext)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:45]: Can we add an `assertIsForeground: false`. just to be explicit this isn't asserting the constructor is created there? I can never remember what the default is.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs [hunk 5]
```diff
@@ -66,21 +65,35 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
             _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
-        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+        private async System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
             ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
         {
+            // This logic preserves teh previous behavior we had with IForegroundNotificationService.
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:71]:
```suggestion
            // This logic preserves the previous behavior we had with IForegroundNotificationService.
```

**Review by jasonmalinowski — APPROVED** [review]
This is good from my perspective, and happy to see this thing disappear, as I never liked it nor convinced it was a terribly good thing to have. I'd wait until @sharwell signs off though if there's any concerns he has.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:45]: the default is `:false`.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:89]: Done.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs:45]: added.!

**Comment by CyrusNajmabadi** [comment]
@sharwell can you ptal. i've integrated your changes into this.

**Pull request merged**